sig-scalability/configs-and-limits/faq.md

```diff
@@ -62,6 +62,31 @@ summarized as:
 - if you can't keep your object size below 100kB, reach out to SIG
   Scalability and discuss the usecase to see how we can make it performant
 
+### How should we code client applications to improve scalability?
+
+As noted above, LIST requests can be particularly expensive. So when working with lists
+that may have more than a few thousand elements, consider these guidelines:
+
+1. When defining a new resource type (new CRD) consider expected numbers 
+of objects that will exist (numbers of CRs).  See guidelines 
+[here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/95-custom-resource-definitions#scale-targets-for-ga).
+1. If your code needs to hold an up-to-date list of objects in memory,
+avoid repeated LIST calls if possible.  Instead consider using the
+`Informer` classes that are provided in most Kubernetes client 
+libraries.   Informers automatically combine LIST and WATCH functionality
+to efficiently maintain an in-memory collection.
+1. If `Informer`s don't suit your needs, try to use the API Server cache 
+when LISTing.  To use the cache you must supply a `ResourceVersion`. 
```

Comment:
Logic here is unfortunately much more complex - if you specify `resourceVersion=0` it will be served from cache. If you specify `resourceVersion=X` and also some limit - it will be served from etcd, but if you don't specify limit - it will be served from cache...

Reply:
to be more specific https://github.com/kubernetes/kubernetes/blob/6dd8b86124e7e4f36f4a3d1939b6ee616ad40c58/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L619-L633


it will be nice if somebody created a blog post or even do a kubecon talk about this, "how to efficiently use the Kubernetes apiserver cache" or "how not to kill your cluster using List"  :smile: 

I don't know how many escalations and bugs I've seen because of this problem of listing from etcd ...

Reply:
💯

Please also note that the decision of when a LIST is delegated to ’etcd’ _might_ change in the coming release(s).
Note: https://github.com/kubernetes/kubernetes/pull/108392

I'm also trying to make the function that decides this, exportable: https://github.com/kubernetes/kubernetes/pull/111281

Reply:
Yeah, those are all good points.  I'm undecided between how much detail to give here, versus keeping the advice as simple as possible.  (Because giving all the detail can be intimidating for first time readers).

Also, re your point re limits @marseel , that's why I did the suggestions as an _ordered_ list, and put "use cache" higher than "use chunking".  

Any thoughts @marseel , @aojea , and @MadhavJivrajani  on how detailed the advice should be?

Reply:
Awaiting your thoughts on whether this FAQ should be detailed or consise, please .. @marseel , @aojea

Reply:
this is a sig-scalability call, but IMHO if you don't explain all the details people will think that using `ResourceVersion` will use the cache always  ... and is not  the reality

Reply:
I would prefer simplification when we recommend setting `ResourceVersion=0` to use cache, which is always true.

Reply:
I'm +1 on saying `ResourceVersion=0`, but that's not enough. We need to say the core point.
So something like:

If you don't need strong consistency, set `ResourceVersion=0` to have the list served from cache.

Reply:
Nice, I like that suggestion @wojtek-t.   I think it gives the necessary context to the simplification that @marseel suggested above. I'm thinking about whether I should also include a definition or explaination of what is meant by "strong consistency".  Will see if I can come up with anything and, if I can, will include it in update to PR.